tty/pc-tty/ttypc.c

```diff
@@ -1,204 +1,184 @@
+			if ((msg.i.attr.type == atPollStatus) && (msg.i.io.oid.id < NVTS))
+				msg.o.attr.val = ttypc_vt_pollstatus(ttypc->vts + msg.i.io.oid.id);
 			else
 				msg.o.attr.val = -EINVAL;
 			break;
+
 		case mtDevCtl:
-			ttypc_ioctl(port, &msg);
+			idata = ioctl_unpack(&msg, &req, &id);
+			if (id < NVTS) {
+				err = ttypc_vt_ioctl(ttypc->vts + id, ioctl_getSenderPid(&msg), req, idata, &odata);
+			}
+			else {
+				odata = NULL;
+				err = -EINVAL;
+			}
+			ioctl_setResponse(&msg, req, err, odata);
 			break;
 		}
 
-		msgRespond(port, &msg, rid);
+		msgRespond(ttypc->port, &msg, rid);
 	}
 }
 
 
 int main(void)
 {
-	void *base = (void *)0x60;
-	unsigned int n = 1;
-	uint32_t port;
+	unsigned int i, j, init = 0;
+	char name[10];
+	oid_t oid;
+	int err;
 
-	printf("pc-tty: Initializing VGA VT220 terminal emulator %s\n", "");
+	/* Initialize driver */
+	memset(&ttypc_common, 0, sizeof(ttypc_t));
 
-	_ttypc_init(base, n);
+	do {
+		if ((err = portCreate(&ttypc_common.port)) < 0)
+			break;
+		init++;
 
-	/* Register port in the namespace */
-	portCreate(&port);
-	if (portRegister(port, "/dev/tty0", &ttypc_common.virtuals[0].oid) < 0) {
-		printf("Can't register port %d\n", port);
-		return -1;
-	}
+		if ((err = mutexCreate(&ttypc_common.lock)) < 0)
+			break;
+		init++;
 
-	/* Run threads */
-	beginthread(poolthr, 1, &ttypc_common.poolthr_stack, sizeof(ttypc_common.poolthr_stack), (void *)port);
-	poolthr((void *)port);
+		/* Initialize and register virtual terminals */
+		for (i = 0; i < NVTS; i++) {
+			snprintf(name, sizeof(name), "/dev/tty%u", i);
+			oid.port = ttypc_common.port;
+			oid.id = i;
+
+			if ((err = portRegister(oid.port, name, &oid)) < 0)
+				break;
+
+			if ((err = ttypc_vt_init(&ttypc_common, _PAGE_SIZE, ttypc_common.vts + i)) < 0)
+				break;
+		}
+		if (err < 0)
+			break;
+		printf("pc-tty: Initialized %u virtual terminals\n", NVTS);
+		init++;
+
+		/* Initialize VGA */
+		if ((err = ttypc_vga_init(&ttypc_common)) < 0)
+			break;
+		printf("pc-tty: Initialized VGA display\n");
+
+		/* Set active virtual terminal */
+		ttypc_common.vt = ttypc_common.vts;
+		ttypc_common.vt->vram = ttypc_common.vga;
+
+		/* Initialize cursor */
+		_ttypc_vga_getcursor(ttypc_common.vt);
+		/* Set default cursor color */
+		memsetw(ttypc_common.vt->vram + ttypc_common.vt->cpos, FG_LIGHTGREY << 8, ttypc_common.vt->rows * ttypc_common.vt->cols - ttypc_common.vt->cpos);
+		init++;
+
+		/* Initialize KBD */
+		if (ttypc_kbd_configure(&ttypc_common)) {
+			if ((err = ttypc_kbd_init(&ttypc_common)) < 0)
+				break;
+			ttypc_common.ktype = KBD_PS2;
+		}
+		else {
+			if ((err = ttypc_bioskbd_init(&ttypc_common)) < 0)
+				break;
+			ttypc_common.ktype = KBD_BIOS;
+		}
+		printf("pc-tty: Initialized %s keyboard\n", (ttypc_common.ktype) ? "PS/2" : "BIOS");
+		init++;
+
+		/* Start poolthread */
+		if ((err = beginthread(ttypc_poolthr, 1, &ttypc_common.pstack, sizeof(ttypc_common.pstack), (void *)&ttypc_common)) < 0)
```

Comment:
ttypc_common.pstack is already a pointer

Reply:
Fixed.